sdk/tables/Azure.Data.Tables/src/TableClient.cs

```diff
@@ -1155,20 +1236,179 @@ public virtual Response SetAccessPolicy(IEnumerable<SignedIdentifier> tableAcl,
+        /// <param name="batchOperations">The <see cref="IEnumerable{T}"/> containing the <see cref="BatchItem"/>s to submit to the service.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
+        /// <returns><see cref="Response{T}"/> containing a <see cref="TableBatchResponse"/>.</returns>
+        /// <exception cref="RequestFailedException"/> if the batch transaction fails./>
+        /// <exception cref="InvalidOperationException"/> if the batch has been previously submitted.
+        public virtual async Task<Response<TableBatchResponse>> SubmitTransactionAsync(
+            IEnumerable<BatchItem> batchOperations,
+            CancellationToken cancellationToken = default) =>
+            await SubmitTransactionInternalAsync(batchOperations, _batchGuid ?? Guid.NewGuid(), _changesetGuid ?? Guid.NewGuid(), true, cancellationToken)
+                .ConfigureAwait(false);
+
+        /// <summary>
+        /// Submits the batch transaction to the service for execution.
+        /// The sub-operations contained in the batch will either succeed or fail together as a transaction.
+        /// </summary>
+        /// <param name="batchOperations">The <see cref="IEnumerable{T}"/> containing the <see cref="BatchItem"/>s to submit to the service.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
+        /// <returns><see cref="Response{T}"/> containing a <see cref="TableBatchResponse"/>.</returns>
+        /// <exception cref="RequestFailedException"/> if the batch transaction fails./>
+        /// <exception cref="InvalidOperationException"/> if the batch has been previously submitted.
+        public virtual Response<TableBatchResponse> SubmitTransaction(IEnumerable<BatchItem> batchOperations, CancellationToken cancellationToken = default) =>
+            SubmitTransactionInternalAsync(batchOperations, _batchGuid ?? Guid.NewGuid(), _changesetGuid ?? Guid.NewGuid(), false, cancellationToken).EnsureCompleted();
+
+        internal virtual async Task<Response<TableBatchResponse>> SubmitTransactionInternalAsync(
+            IEnumerable<BatchItem> transactionalBatch,
+            Guid batchId,
+            Guid changesetId,
+            bool async,
+            CancellationToken cancellationToken = default)
         {
-            return new TableTransactionalBatch(Name, partitionKey, _tableOperations, _defaultQueryOptions.Format.Value, _isCosmosEndpoint);
+            using DiagnosticScope scope = _diagnostics.CreateScope($"{nameof(TableClient)}.{nameof(SubmitTransaction)}");
+            scope.Start();
+            try
+            {
+                Argument.AssertNotNull(transactionalBatch, nameof(transactionalBatch));
+                List<BatchItem> batchItems = transactionalBatch.ToList();
+                if (!batchItems.Any())
```

Comment:
You're enumerating multiple times; since the `IEnumerable<BatchItem>` is provided by the caller, there's no guarantee that it is safe to do so.  Any chance that we can use `BuildChangeSet` to tell us if it found any items and throw if not?

Reply:
`batchItems` is a materialized list. `List<BatchItem> batchItems = transactionalBatch.ToList();`

Reply:
d'oh.   One line above and I overlooked that.